### PR TITLE
Mapping state deserializer

### PR DESCRIPTION
**Pull request by sebastianklose**
class for de-serializing mapping state stored in the new serialization format [RFC 0021](https://github.com/sebastianklose/rfcs/blob/serialization_format/text/0021-serialization-format.md).

Not used yet. Will be hooked up in a separate PR that switches all logic to the new serialization format at once.

### CODE REVIEW

File: cartographer/io/mapping_state_deserializer_test.cc [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+// This test checks, if the serialization works.
+TEST(MappingStateDeserializerTest, WorksOnGoldenTextStream) {
+  // Load text proto into in_memory_reader.
+  std::unique_ptr<InMemoryProtoStreamReader> reader =
+      CreateInMemoryReaderFromTextProtos();
+
+  io::MappingStateDeserializer deserializer(reader.get());
+
+  EXPECT_TRUE(google::protobuf::util::MessageDifferencer::Equals(
```
Comment (pifon2a) [inline cartographer/io/mapping_state_deserializer_test.cc]: please, add `using google::protobuf::util::MessageDifferencer`

File: cartographer/io/mapping_state_deserializer_test.cc [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+
+template <typename T>
+T ProtoFromStringOrDie(const std::string& proto_string) {
+  T msg;
+  CHECK(google::protobuf::TextFormat::ParseFromString(proto_string, &msg));
+  return msg;
+}
+
+template <typename T>
+std::unique_ptr<T> CreateProtoUPtrFromString(const std::string& proto_string) {
```
Comment (gaschler) [inline cartographer/io/mapping_state_deserializer_test.cc]: Rename to CreateProtoUniquePtrFromString

File: cartographer/io/mapping_state_deserializer_test.cc [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+using common::make_unique;
+using google::protobuf::Message;
+using mapping::proto::AllTrajectoryBuilderOptions;
+using mapping::proto::PoseGraph;
+using mapping::proto::SerializationHeader;
+using mapping::proto::SerializedData;
+using ::testing::Eq;
+using ::testing::Not;
+
+void ProtoFromStringOrDie(const std::string& proto_string,
```
Comment (pifon2a) [inline cartographer/io/mapping_state_deserializer_test.cc]: I am not sure you really need
```
void ProtoFromStringOrDie(const std::string& proto_string,  google::protobuf::Message* msg)
std::unique_ptr<T> CreateProtoUPtrFromString(const std::string& proto_string) 
```
see the comments below.

File: cartographer/io/mapping_state_deserializer_test.cc [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+  CHECK(google::protobuf::TextFormat::ParseFromString(proto_string, &msg));
+  return msg;
+}
+
+template <typename T>
+std::unique_ptr<T> CreateProtoUPtrFromString(const std::string& proto_string) {
+  return make_unique<T>(ProtoFromStringOrDie<T>(proto_string));
+}
+
+// open text file with a set of text-protos that mimic a pbstream.
```
Comment (gaschler) [inline cartographer/io/mapping_state_deserializer_test.cc]: clarify "open text file", it is more like protos in memory.

File: cartographer/io/mapping_state_deserializer_test.cc [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+                       proto_queue.back().get());
+
+  proto_queue.push(make_unique<mapping::proto::SerializedData>());
+  ProtoFromStringOrDie(testing::kLandmarkDataProtoString,
+                       proto_queue.back().get());
+
+  return make_unique<InMemoryProtoStreamReader>(std::move(proto_queue));
+}
+
+// This test checks, if the serialization works.
```
Comment (gaschler) [inline cartographer/io/mapping_state_deserializer_test.cc]: drop comma

File: cartographer/io/mapping_state_deserializer_test.cc [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+  EXPECT_TRUE(google::protobuf::util::MessageDifferencer::Equals(
+      deserializer.header(), ProtoFromStringOrDie<SerializationHeader>(
+                                 testing::kSerializationHeaderProtoString)));
+
+  EXPECT_TRUE(google::protobuf::util::MessageDifferencer::Equals(
+      deserializer.pose_graph(),
+      ProtoFromStringOrDie<SerializedData>(testing::kPoseGraphProtoString)
+          .pose_graph()));
+
+  EXPECT_TRUE(google::protobuf::util::MessageDifferencer::Equals(
```
Comment (gaschler) [inline cartographer/io/mapping_state_deserializer_test.cc]: You could use EXPECT_EQ and a matcher.
Then, a failed test would print the two arguments.

File: cartographer/io/mapping_state_deserializer.h [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+  const mapping::proto::AllTrajectoryBuilderOptions&
+  all_trajectory_builder_options() {
+    return all_trajectory_builder_options_.all_trajectory_builder_options();
+  }
+
+  // Reads the next `SerializedData` message of the ProtoStream into `data`.
+  // Returns `true` if the message was successfully read or `false` in case
+  // there are no-more messages or an error occurred.
+  bool GetNextSerializedData(mapping::proto::SerializedData* data);
```
Comment (gaschler) [inline cartographer/io/mapping_state_deserializer.h]: Rename to Read*

File: cartographer/io/mapping_state_deserializer.h [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+#include "cartographer/mapping/proto/pose_graph.pb.h"
+#include "cartographer/mapping/proto/serialization.pb.h"
+#include "cartographer/mapping/proto/trajectory_builder_options.pb.h"
+
+namespace cartographer {
+namespace io {
+
+// Class to help deserializing a previously serialized mapping state from a
+// proto stream, abstracting away the format parsing logic.
+class MappingStateDeserializer {
```
Comment (gaschler) [inline cartographer/io/mapping_state_deserializer.h]: Will all deserialization go through this later? Or will it be used by the ground truth and ros tools that only need the pose graph?
Rename. Perhaps PoseGraph/ProtoStreamDeserializer?

File: cartographer/io/mapping_state_deserializer.cc [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+ * limitations under the License.
+ */
+
+#include "cartographer/io/mapping_state_deserializer.h"
+#include "cartographer/io/internal/mapping_state_serialization.h"
+#include "glog/logging.h"
+
+namespace cartographer {
+namespace io {
+
```
Comment (gaschler) [inline cartographer/io/mapping_state_deserializer.cc]: nit: move empty line after namespace

File: cartographer/io/internal/testing/serialized_mapping_state_text_proto.h [hunk 10]
```diff
@@ -0,0 +1,211 @@
+#ifndef CARTOGRAPHER_IO_INTERNAL_TESTING_SERIALIZED_TEST_TEXT_PROTO_H_
+#define CARTOGRAPHER_IO_INTERNAL_TESTING_SERIALIZED_TEST_TEXT_PROTO_H_
+
+#include <string>
+
+namespace cartographer {
+namespace io {
+namespace testing {
+
+static const std::string kSerializationHeaderProtoString = R"(
```
[the remark this answers is not in the document]
Comment (sebastianklose) [reply cartographer/io/internal/testing/serialized_mapping_state_text_proto.h]: done

File: cartographer/io/internal/testing/serialized_mapping_state_text_proto.h [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+      trajectory_id: 0
+      submap_index: 0
+    }
+    node_id {
+      trajectory_id: 0
+      node_index: 0
+    }
+    relative_pose {
+      translation {
+        x: 0.0
```
Comment (gaschler) [inline cartographer/io/internal/testing/serialized_mapping_state_text_proto.h]: choose non-default values?

File: cartographer/io/mapping_state_deserializer.h [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+
+#include "cartographer/io/proto_stream_interface.h"
+#include "cartographer/mapping/proto/pose_graph.pb.h"
+#include "cartographer/mapping/proto/serialization.pb.h"
+#include "cartographer/mapping/proto/trajectory_builder_options.pb.h"
+
+namespace cartographer {
+namespace io {
+
+// Class to help deserializing a previously serialized mapping state from a
```
[the remark this answers is not in the document]
Comment (sebastianklose) [reply cartographer/io/mapping_state_deserializer.h]: done

File: cartographer/io/mapping_state_deserializer.h (hunk 8 above)
Comment (sebastianklose) [reply cartographer/io/mapping_state_deserializer.h]: Renamed.

File: cartographer/io/mapping_state_deserializer_test.cc (hunk 2 above)
Comment (sebastianklose) [reply cartographer/io/mapping_state_deserializer_test.cc]: renamed to `ProtoUPtrFromStringOrDie`

File: cartographer/io/mapping_state_deserializer_test.cc (hunk 6 above)
Comment (sebastianklose) [reply cartographer/io/mapping_state_deserializer_test.cc]: added TODO

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user pifon2a. Merge queue now has a length of 2.

**Pull request merged**